### PR TITLE
Issue #5681 - clearer warning on JVM Arg / System Property use

### LINKED ISSUE #5681: Unrecognized jetty-home/start.jar command line option not reported clearly
**jetty-10.0.x**

If I use an unknown command option to start the distro like so:
[941] java -jar ../jetty-home/start.jar --foople

Jetty simply starts. I would expect at least an "unknown option" type warning, maybe even not starting.

### PR DESCRIPTION
**Pull request by joakime**
Fixes #5681 Unknown distro command ignored

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-start/src/main/java/org/eclipse/jetty/start/Main.java [hunk 1]
```diff
@@ -502,6 +502,10 @@ public void run()
         if (args.hasJvmArgs() || args.hasSystemProperties())
         {
             StartLog.warn("System properties and/or JVM args set.  Consider using --dry-run or --exec");
+            if (args.hasSystemProperties())
+                args.getSystemProperties().forEach((k, v) -> StartLog.warn("  Detected JVM System Property: %s=%s", k, System.getProperty(k)));
+            if (args.hasJvmArgs())
+                args.getJvmArgs().forEach((jvmArg) -> StartLog.warn("  Detected JVM Arg: %s", jvmArg));
```
Comment (joakime) [inline jetty-start/src/main/java/org/eclipse/jetty/start/Main.java]: Other language options.

```
  Detected as JVM System Property: %s=%s
  Detected as JVM Arg: %s
```

```
  Detected possible JVM System Property: %s=%s
  Detected possible JVM Arg: %s
```

Comment (joakime) [reply jetty-start/src/main/java/org/eclipse/jetty/start/Main.java]: From @sbordet 

```
WARN  : Unknown Arguments detected.  Consider using --dry-run or --exec
WARN  :   Argument: -Dzed.key=0.value (interpreted as a system property, from <command-line>)
WARN  :   Argument: --foople (interpreted as a JVM argument)
```

**Comment by joakime** [comment]
@sbordet new output syntax committed.

Example output:

```
WARN  : Unknown Arguments detected.  Consider using --dry-run or --exec
WARN  :   Argument: -Dzed.key=0.value (interpreted as a System property, from <command-line>)
WARN  :   Argument: --zed-0-zed (interpreted as a JVM argument, from /home/joakim/code/jetty/jetty.project-10.0.x/jetty-start/target/tests/test-base-example-unknown/start.d/zed.ini)
WARN  :   Argument: --foople (interpreted as a JVM argument, from <command-line>)
```

File: jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java [hunk 2]
```diff
@@ -1331,10 +1357,12 @@ public void parse(final String rawarg, String source)
...
         if (arg.startsWith("-"))
         {
+            StartLog.debug("Unrecognized Arg: %s (from %s)", arg, source);
+
             // Only add non-duplicates
-            if (!jvmArgs.contains(arg))
+            if (!jvmArgSources.containsKey(arg))
             {
-                jvmArgs.add(arg);
+                jvmArgSources.put(arg, source);
```
Comment (sbordet) [inline jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java]: Use `putIfAbsent()` here.

File: jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java [hunk 3]
```diff
@@ -331,21 +331,23 @@ public void dumpEnvironment(PrintStream out)
...
+        jvmArgSources.forEach((key, sourceRef) ->
+            {
+                String value = System.getProperty(key);
+                String source = StartLog.isDebugEnabled() ? '(' + sourceRef + ')' : "";
+                if (value != null)
+                    out.printf(" %s = %s %s%n", key, value, source);
+                else
+                    out.printf(" %s %s%n", key, source);
+            }
+        );
```
Comment (sbordet) [inline jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java]: Weird braces and parens alignment, can you make it normal?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java (hunk 2 above)
Comment (joakime) [reply jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java]: Disagree.
This arg can be overridden by a the same arg later on, which could have a different source.
The last source is what's used, so we replace each time to know the eventually used source.

File: jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java (hunk 3 above)
Comment (joakime) [reply jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java]: This passes checkstyle.
I'll give it whirl.

File: jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java (hunk 2 above)
Comment (sbordet) [reply jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java]: I don't understand? There is a `!containsKey()` followed by `put()` so it's equivalent to `putIfAbsent()`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java (hunk 2 above)
Comment (joakime) [reply jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java]: You are correct that `putIfAbsent` is an appropriate replacement for what you reviewed.
I disagree that it should be used at all, the check `!containsKey()` shouldn't even be there too.
I pushed an update.

File: jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java (hunk 3 above)
Comment (joakime) [reply jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java]: Done.

**Review by sbordet — APPROVED** [review]

**Pull request merged**